Upgrade Material Dialogs

mikehardy opened the pull request:
## Pull Request template

Please, go through these checks before you submit a PR.

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

## Purpose / Description
The Material Dialogs dependency was the next in line holding us back from forward compatibility with APIs > 26, upgrading it lets us go to API27

Also the deprecation warnings were log spam

## Fixes
The library has documentation to the effect that there were tons of bug fixes and forward support between our version and 0.9.6.0

## Approach

You will want to review this commit by commit, probably side by side. Each one is compartmentalized (and individually releasable, at that).

The only big one was the move to RecyclerView for saved searches. My first UI change, exciting.

-  prep codebase with commit for Intent.FLAG_ACTIVITY_CLEAR_TASK
-  prep codebase with commit for RecyclerView/DividerItemDecoration
-  Alter the dependency and move APIs to 27 in gradle, clean up file
-  Use library example custom recycler view adapter to implement SavedSearches
-  Codacy cleanup pass (no logic changes)
-  Deprecation cleanup pass (mechanical change to listener style, no logic)

## How Has This Been Tested?

Used it in emulators from API15 to API28

## Learning (optional, can help others)

The framework implemented a RecyclerView DividerItemDecoration, which is nice.

MaterialDialogs is at 0.9.6.0 now but is in alpha for a 2.0, something for far future...

Java 1.8 Lambdas (which are valid for all target APIs, I checked!) are pretty clean, syntactically. So are method references (also valid for all target APIs)

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.java:
```diff
@@ -190,6 +168,8 @@ public void onSelection(MaterialDialog materialDialog, View view, int which,
                                         ((DeckPicker) getActivity())
                                                 .showDatabaseErrorDialog(DIALOG_NEW_COLLECTION);
+                                    default:
+                                        throw new RuntimeException("Unknown dialog selection: " + mRepairValues[which]);
```
I'm following codebase convention here.

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:
```diff
@@ -1473,11 +1477,11 @@ public void onScroll(AbsListView view, int firstVisibleItem, int visibleItemCoun
                 // Note: max value of lastVisibleItem is totalItemCount, so need to subtract 1
                 String lastAns = getCards().get(lastVisibleItem - 1).get("answer");
-                if (firstAns != null && firstAns.equals("") || lastAns != null && lastAns.equals("")) {
+                if (firstAns != null && "".equals(firstAns) || lastAns != null && lastAns.equals("")) {
```
Yoda'd one I did, forget the second it is clear. Leave this for now I will.

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/ImportDialog.java:
```diff
@@ -70,17 +69,8 @@ public MaterialDialog onCreateDialog(Bundle savedInstanceState) {
-                            }
-                        })
+                        .onPositive((dialog, which) -> ((ImportDialogListener) getActivity()).showImportDialog(DIALOG_IMPORT_SELECT))
+                        .onNeutral((dialog, which) -> dismissAllDialogFragments())
```
This should be .onNegative not .onNeutral, fixing...

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:
```diff
@@ -1477,11 +1477,11 @@ public void onScroll(AbsListView view, int firstVisibleItem, int visibleItemCoun
                 // Note: max value of lastVisibleItem is totalItemCount, so need to subtract 1
                 String lastAns = getCards().get(lastVisibleItem - 1).get("answer");
-                if (firstAns != null && firstAns.equals("") || lastAns != null && lastAns.equals("")) {
+                if (firstAns != null && "".equals(firstAns) || lastAns != null && lastAns.equals("")) {
```
Remove the null check and do yoda on the second half of the `if` expression

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/CardBrowserMySearchesDialog.java:
```diff
@@ -30,19 +22,19 @@
     private String mCurrentSearchTerms;
 
+    public interface MySearchesDialogListener {
+        public void OnSelection(String searchName);
```
`public` is redundant in an interface

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.java:
```diff
@@ -75,18 +75,9 @@ public MaterialDialog onCreateDialog(Bundle savedInstanceState) {
+                        .onNegative((dialog, which) -> ((DeckPicker) getActivity()).exit())
                         .show();
 
             case DIALOG_DB_ERROR:
```
Can you please add braces to these `case` statements? Any case statement that declares a variable should have curly braces to limit the scope.

mikehardy replied:
Sure - trying to balance patch size for review and not feel like a bull in a china shop as I work through this old code, but totally agreed

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.java:
```diff
@@ -99,24 +90,13 @@ public void onNegative(MaterialDialog dialog) {
-                            public void onNeutral(MaterialDialog dialog) {
-                                ((DeckPicker) getActivity()).exit();
-                            }
+                        .onPositive((dialog1, which) -> ((DeckPicker) getActivity())
```
Wow java lambdas feel so crude compared to C++11... It seems in the future [you'll be allowed](http://openjdk.java.net/jeps/302) to use underscores for unused variables. Until then maybe name these all `inner_dialog` instead of `dialogxy`

mikehardy replied:
Lambdas are actually completely new to me (apart from Scheme in college in the 90s (!!)) so I was thinking they were all spiffy ;-). Really does snug up the code and the method reference in the anonymous Comparator was a revelation. These were generated from Android Studio's auto-convert feature thus the odd naming. But I can scan through and fix up sure

timrae commented:
Can you please rebase and also squash fixes to make something a bit more coherent after addressing the minor comments?

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.java:
```diff
@@ -160,58 +137,48 @@ public void onNeutral(MaterialDialog dialog) {
+                                    ((DeckPicker) getActivity()).showDatabaseErrorDialog(DIALOG_NEW_COLLECTION);
+                                    return;
+                                default:
+                                    throw new RuntimeException("Unknown dialog selection: " + mRepairValues[which]);
```
As noted before - following codebase practice here

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:
```diff
@@ -2186,7 +2170,7 @@ public void onExtendStudyLimits() {
      * There was pre-honeycomb override here, removed now, just instantiate regular callback
      */
-    Snackbar.Callback mSnackbarShowHideCallback = new Snackbar.Callback();
+    private Snackbar.Callback mSnackbarShowHideCallback = new Snackbar.Callback();
```
Can't fix everything without generating a ridiculous diff - might already be too much for this round...

mikehardy commented:
Okay I rebased, and I squashed the cleanups into one last commit.

It's hard to start doing codacy stuff without touching lots of things rapidly and I fixed quite a few but didn't attempt to fix everything or the diff would be crazy

I still like the others separate for review (split screen / no whitespace makes them easy for me to see) but I can reshape the commit however if you like.

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java:
```diff
@@ -1095,12 +1087,13 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
 
         switch (requestCode) {
-            case REQUEST_ADD:
+            case REQUEST_ADD: {
```
You don't need braces when assigning variables, just when *declaring* them

timrae commented:
Yeah this is a massive diff, I can't be bothered to go through it line by line, let's just test it well
